lms/djangoapps/course_structure_api/v0/views.py

```diff
@@ -290,3 +296,299 @@ class CourseGradingPolicy(CourseViewMixin, ListAPIView):
+                if children_requested:
+                    block_value["children"] = [unicode(child.location) for child in children]
+
+            # graded
+            # If graded information is requested, include both
+            # (1) whether this block is graded and
+            # (2) whether this block or any of the block's descendants is graded.
+            if graded_requested:
+                block_value["graded"] = block.graded
+                block_value["graded_descendant"] = (
+                    block.graded or
+                    any(
+                        blocks.get(unicode(child.location), {}).get("graded_descendant", False)
+                        for child in children
+                    )
+                )
+                block_value["format"] = getattr(block, 'format', None)
+
+            # responsive UI
+            # If responsive UI information is requested, include whether this block is tagged with having
+            # responsive UI support.
+            if responsive_ui_requested:
+                block_value["responsive_ui"] = (
+                    block.has_responsive_ui if hasattr(block, 'has_responsive_ui') else False
+                )
+
+            # block count
+            # For all the block types that are requested to be counted, include the count of
+            # that block type as aggregated from the block's descendants.
+            for b_type in block_count_requested:
+                block_value.setdefault("block_count", {})[b_type] = (
+                    sum(
+                        blocks.get(unicode(child.location), {}).get("block_count", {}).get(b_type, 0)
+                        for child in children
+                    ) +
+                    (1 if b_type == block_type else 0)
+                )
+
+            # block data
+            # If the data for this block's type is requested, and the block supports the 'student_view_json' method,
+            # add the response from the 'student_view_json" method as the data for the block.
+            # TODO Call student_view_json instead, but the block is a descriptor instead of the module.
+            if block_type in block_json_requested:
+                if getattr(block, 'get_json', None):
+                    block_value["data"] = block.get_json(
+                        context=block_json_requested[block_type]
+                    )
+
+        # start the recursion with the course at block_depth 0
+        start_block = course
+        response["root"] = unicode(start_block.location)
+        recurse_blocks_nav(start_block, block_depth=0, descendants_of_parent=[])
+
+        # return response
+        return Response(response)
+
+
+def create_module(descriptor, course_id, request):
+    """
+    Factory method for creating and binding a module for the given descriptor.
+    """
+    field_data_cache = FieldDataCache.cache_for_descriptor_descendents(
+        course_id, request.user, descriptor, depth=0,
+    )
```

Comment:
I would like to see tests using `assertNumQueries` and `check_mongo_calls`. In particular, I'm worried about what look to be repeated calls to `cache_for_descriptor_descendents` with a depth of 0, and that that will end up fetching student state for each XBlock as a single query, rather than loading them all together.

If you find that it really is better to do the queries that way, then you can just use the `FieldDataCache` constructor directly, and pass in the single descriptor that you're loading, rather than using `cache_for_descriptor_descendents`.